KsApi/extensions/MockGraphQLClient.swift

```diff
@@ -0,0 +1,116 @@
+import Apollo
+import Foundation
+import ReactiveSwift
+
+class MockGraphQLClient: ApolloClientType {
+  // MARK: - Base Properties
+
+  var client: ApolloClient {
+    let url = URL(string: "https://kickstarter.com")!
+
+    return ApolloClient(url: url)
+  }
+
+  static let shared = MockGraphQLClient()
+
+  // MARK: Public functions
+
+  /// Placeholder implementation because protocol definition used in `Service`
+  public func fetch<Query: GraphQLQuery>(query _: Query) -> SignalProducer<Query.Data, ErrorEnvelope> {
+    .init(value: Query.Data(unsafeResultMap: [:]))
+  }
+
+  /// Placeholder implementation because protocol definition used in `Service`
+  public func perform<Mutation: GraphQLMutation>(
+    mutation _: Mutation
+  ) -> SignalProducer<Mutation.Data, ErrorEnvelope> {
+    .init(value: Mutation.Data(unsafeResultMap: [:]))
+  }
+}
+
+/** Implementation of optional `fetch` and `perform` with `result` useful for mocking data.
+ */
+extension ApolloClientType {
+  public func fetchWithResult<Query: GraphQLQuery, Data: Decodable>(
+    query _: Query,
+    result: Result<Data, ErrorEnvelope>?
+  ) -> SignalProducer<Data, ErrorEnvelope> {
+    producer(for: result)
+  }
+
+  public func performWithResult<Mutation: GraphQLMutation, Data: Decodable>(
+    mutation _: Mutation,
+    result: Result<Data, ErrorEnvelope>?
+  ) -> SignalProducer<Data, ErrorEnvelope> {
+    return producer(for: result)
+  }
+
+  public func dataFromProducer<Data: Decodable>(_ producer: SignalProducer<Data, ErrorEnvelope>) -> Data? {
+    switch producer.first() {
+    case let .success(data):
+      return data
+    default:
+      return nil
+    }
+  }
+
+  public func errorFromProducer<Data: Decodable>(_ producer: SignalProducer<Data, ErrorEnvelope>)
+    -> ErrorEnvelope? {
+    switch producer.first() {
+    case let .failure(errorEnvelope):
+      return errorEnvelope
+    default:
+      return nil
+    }
+  }
+}
+
+private func producer<T, E>(for property: Result<T, E>?) -> SignalProducer<T, E> {
+  guard let result = property else { return .empty }
+  switch result {
+  case let .success(value): return .init(value: value)
+  case let .failure(error): return .init(error: error)
+  }
+}
+
+private extension Result {
+  var value: Success? {
+    switch self {
+    case let .success(value): return value
+    case .failure: return nil
+    }
+  }
+
+  var error: Failure? {
+    switch self {
+    case .success: return nil
+    case let .failure(error): return error
+    }
+  }
+}
+
+// TODO: Is it possible in the future to return a type of `Query.Data` instead of `Decodable` using this?
```

Comment:
If we aren't using this, we should delete it. After working on this PR do you think we can incorporate this within the testing implementation we have here?

Reply:
No I tried a variety of ways. 
It's just that `Query.Data` and `Mutation.Data` are used internally by GraphAPI to create the queries and mutations.
They all contain their own implementation of `Data` which then may also contain it's own dependencies with their own implementation of `Data`. (ie. `WatchProjectMutation` has a `Data` object, which has a `WatchProject` struct that has it's own `Data` object)

So its not possibly to re-create an entirely new GraphAPI schema just for testing.

Reply:
Actually this file was being used in a different context inside `ApolloInterceptorTests`. Left it in, removed comment.